### ISSUE TITLE
Production Checklist: Add OS settings

### ISSUE THREAD

**Issue by lnhsingh** [cockroachdb/docs#4209]
@drewdeally: for OS settings, I will work on a doc which covers all OS setting recommendations and implications , tested and have validated by eng team. Such as the following:

```
echo "vm.overcommit_memory = 2" >> /etc/sysctl.conf
echo "vm.overcommit_ratio = 99" >> /etc/sysctl.conf
echo "vm.swappiness = 0" >> /etc/sysctl.conf # Or set to 0 to disable swap. Overcommit mixed with this can have swap implications
echo "vm.max_map_count=1000000000" >> /etc/sysctl.conf
echo "vm.min_free_kbytes=500000" >> /etc/sysctl.conf
echo "net.core.somaxconn=1024" >> /etc/sysctl.conf
echo never > /sys/kernel/mm/transparent_hugepage/enabled
echo never > /sys/kernel/mm/transparent_hugepage/defrag
echo "echo never > /sys/kernel/mm/transparent_hugepage/enabled" >> /etc/rc.d/rc.local
echo "echo never > /sys/kernel/mm/transparent_hugepage/defrag" >> /etc/rc.d/rc.local
```

Related to #4153.

**Comment by jseldess** [comment]
@lhirata, was this part of your product checklist work, or are we waiting for details from @drewdeally?

**Comment by drewdeally** [comment]
Oh - please hold off I need to validate these with some content..

**Comment by drewdeally** [comment]
vm.overcommit_memory = 2 >> /etc/sysctl.conf will need to be reevaluated.    setting to 0 seems to be safe.

**Comment by jhatcher9999** [comment]
not sure if this is helpful or not, but Datastax has a lot of these same recommended settings and might have some good general descriptions that we could leverage: https://docs.datastax.com/en/dse/6.7/dse-dev/datastax_enterprise/config/configRecommendedSettings.html

**Comment by drewdeally** [comment]
Agreed, started a document early on https://docs.google.com/document/d/171XnKag1xkIh59MkU8g8_7abg0pQfV2O7nUfTIixssA/edit <https://docs.google.com/document/d/171XnKag1xkIh59MkU8g8_7abg0pQfV2O7nUfTIixssA/edit>
But before adding into docs, we need to test the settings and behavior as it related to CRDB.

@Lauren and I were to pick this up last year.

**Comment by lauren** [comment]
@drewdeally wrong Lauren.

**Comment by drewdeally** [comment]
Sorry --  corrected @lnhsingh

**Comment by jseldess** [comment]
@drewdeally, you closed this issue some time ago. Should this be re-opened. @taroface has taken over this work in the meantime.

**Comment by taroface** [comment]
Reopened per @fabiog1901

**Comment by lancel66** [comment]
Hi, this just came up in a discussion with CRL CEAs. We need to update our live training decks with the latest recommended Linux settings, so this would be helpful. Also, there is some concern that `vm_swappiness=0` does not completely disable swap and has caused issues as noted in the code comments above. It should be probably be replaced with `swapoff -a` + removing the swap device/file. cc @fabiog1901

**Comment by drewdeally** [comment]
currently check the following with customers 

```sudo sysctl -a | egrep 'overcommit_memory|swappiness|overcommit_ratio'
vm.overcommit_memory = 0
vm.overcommit_ratio = 50
vm.swappiness = 0```

it would be great to have updated documents as a reference.

**Comment by lancel66** [comment]
Here is an issue experienced with one of our customers as outlined by Jon S. J. (CEA) related to this:

> overcommit_memory=2     ## Disable OOM
This, combined with a default value of overcommit_ratio = 50 was resulting in an inability for the Pebble cache to allocate new memory and panic() , crashing crdb randomly. My understanding is that this out of memory situation is a risk when cockroach is started with --cache=.25 and --sql-max-memory=.25  which means that it will eventually hit the overcommit_ratio (one being represented as a % and the other as a fraction). When the process attempts to get memory after it has hit the overcommit_ratio, it will be denied but does not get killed by the system. It is left to handle the situation itself, which in crdb’s case is to exit.
I don’t think overcommit_memory=2 is part of the standard documentation so we were trying to figure out where that recommendation came from and whether it should be removed from the configuration and sizing deck/workshop since it could cause some issues.

**Comment by fabiog1901** [comment]
We have a line that says "Disable Linux memory swapping.". we should leave it up to the customer/sysadmin to run the necessary commands to disable swap.

As per `overcommit_memory`, I look forward for our engineering team response. Until then, we should not mention anything in our presentations and docs (docs in fact don't mention anything about it).

**Comment by taroface** [comment]
Hi @fabiog1901, should I take this to mean that we shouldn't document [these settings](https://github.com/cockroachdb/docs/issues/4209#issuecomment-930498065) publicly?

Per the original issue description, there other OS settings that we should be documenting?

**Comment by fabiog1901** [comment]
Personally, I wouldn't - we should leave it up to the customer to know what commands to run on their OS to achieve the desired result. We might show it as an example, but not as a guide. I would like to see a guidance though for sure on the 3 settings, though, pretty much like we have it for swappiness: we don't give the commands to disable it, we just say it in plain English: "Disable Linux memory swapping.".

Ultimately, our PM/ENG for Security should make the call.

**Comment by lancel66** [comment]
> Hi @fabiog1901, should I take this to mean that we shouldn't document [these settings](https://github.com/cockroachdb/docs/issues/4209#issuecomment-930498065) publicly?
> 
> Per the original issue description, there other OS settings that we should be documenting?

I highly recommend that we give examples of what settings to apply on the platforms we support. All of the customers I have dealt with have asked for this and you can find them for any other major database platform.

**Comment by lancel66** [comment]
> Ultimately, our PM/ENG for Security should make the call.

Yes, that is why this issue was filed.

**Comment by taroface** [comment]
@mwang1026 Could you advise on how we might proceed here?

**Comment by jonstjohn** [comment]
Confusion with the `overcommit_memory` in conjunction with an incorrectly chosen `overcommit_ratio` led to frequent (every 6-8 hours) cockroach node crashes for one particular customer. This somewhat obscure issue took a couple of weeks to resolve. 

Although the production checklist in the official docs is great on a high-level, there are quite a few ways that a customer can get it wrong. Ideally, I think we would give some concrete examples of optimal settings for major OSes. We do a pretty good job of this with the file description limit section. Maybe it would be helpful to do something similar for memory settings.

**Comment by taroface** [comment]
As a compromise until we can get PM/ENG guidance on best recommended settings, could we add a callout to the Production Checklist that lists what _not_ to do here? Are there specific combinations of `overcommit_memory`, `overcommit_ratio`, `--cache`, and `--max-sql-memory` that we can warn against using?

**Comment by mwang1026** [comment]
cc @piyush-singh

**Comment by lin-crl** [comment]
From discussion w/ @bobvawter Insights shared:  CC dedicated does a 90% allocation.  for production clusters we may want to reserve more space for metrics collectors, package updates, etc.

**Comment by bschoening** [comment]
MongoDB has similar recommendations to DataStax in their [production-notes](https://www.mongodb.com/docs/manual/administration/production-notes/)

One key issue we hit (which is discussed in the Mongo doc) is RedHat's  vm.force_cgroup_v2_swappiness and their change to how swappiness may be ignored if this isn't correctly set.

**Comment by cdprete** [comment]
Hello.

Are there any updates on this?
I would love to see them officially documented and, therefore, supported as it's in, for example, https://docs.yugabyte.com/preview/deploy/manual-deployment/system-config/#set-ulimits.
Actually, I wonder if `clock-bound` mentioned in there may be helpful as well.

**Comment by dweatherbee** [comment]
@mwang1026 is this something in your wheelhouse? if so, do we have a rec?
